### PR TITLE
Applications: nrf5340_audio: Added audio ablk

### PR DESCRIPTION
**Pull request by koffes**
to be squashed

### CODE REVIEW

File: include/audio_defines.h [hunk 1]
```diff
@@ -11,6 +11,8 @@
 #ifndef _AUDIO_DEFINES_H_
 #define _AUDIO_DEFINES_H_
 
+#include <zephyr/kernel.h>
```
Comment (alexsven) [inline include/audio_defines.h:14]: This is probably not needed, should most likely be types.h or something similar

File: include/audio_defines.h [hunk 2]
```diff
@@ -20,4 +22,105 @@ enum audio_channel {
...
+	/* Number of bits used to carry a sample of size bits_per_sample.
+	 * For example, say we have a 24 bit sample stored in a 32 bit
+	 * word (int32_t), then:
+	 *     bits_per_sample = 24
+	 *     carrier_size    = 32
+	 */
+	uint8_t carrier_size;
+
+	/* A flag indicating if the samples are interleaved or not */
+	enum ablk_interleaved interleaved;
```
Comment (alexsven) [inline include/audio_defines.h:94]: Should this be a bool?

File: include/audio_defines.h [hunk 3]
```diff
@@ -20,4 +22,105 @@ enum audio_channel {
...
+	 * To get the size of each channel, this value must be divided by the number of
+	 * used channels.
+	 */
+	size_t data_size;
+
+	/* Frame length in microseconds.
+	 * This may be aligned with the Bluetooth connection interval.
+	 * Typical values may be 7500 or 10000 us.
+	 */
+	uint32_t frame_len_us;
```
Comment (alexsven) [inline include/audio_defines.h:67]: Should we add number_of_frames? If we know how many frames are in the block, we can calculate the bitrate by using data_size and frame_len_us

**Review by alexsven — CHANGES_REQUESTED** [review]

**Review by asbjornsabo — CHANGES_REQUESTED** [review]
This looks generally good.
It seems quite tied to the LE Audio / ISO specs though - we may want to discuss how tied we want to be.
And we (now or later) need to have a discussion on our audio timing.
Also lots of comments on naming.

File: include/audio_defines.h [hunk 4]
```diff
@@ -20,4 +23,91 @@ enum audio_channel {
...
+	enum type data_type;
+
+	/* Playback length in microseconds.
+	 */
+	uint32_t data_len_us;
+
+	/* The audio sample rate.
+	 * This may typically be 16000, 24000, 44100 or 48000 Hz.
+	 */
+	uint32_t sample_rate_hz;
```
[the remark this answers is not in the document]
Comment (koffes) [reply include/audio_defines.h:48]: I see no reason why we should limit this here. This structure is only to convey information.

File: include/audio_defines.h [hunk 5]
```diff
@@ -20,4 +23,91 @@ enum audio_channel {
...
+	TYPE_LC3
+};
+
+struct audio_metadata {
+	/* Indicates the data type/encoding. */
+	enum type data_type;
+
+	/* Playback length in microseconds.
+	 */
+	uint32_t data_len_us;
```
[the remark this answers is not in the document]
Comment (koffes) [reply include/audio_defines.h:43]: It cannot if it is encoded.

[the remark this answers is not in the document]
Comment (koffes) [reply include/audio_defines.h:43]: We don't have the number of samples

File: include/audio_defines.h [hunk 6]
```diff
@@ -20,4 +23,91 @@ enum audio_channel {
...
+	 */
+	uint8_t bits_per_sample;
+
+	/* Number of bits used to carry a sample of size bits_per_sample.
+	 * For example, say we have a 24 bit sample stored in a 32 bit
+	 * word (int32_t), then:
+	 *     bits_per_sample = 24
+	 *     carrier_size    = 32
+	 */
+	uint8_t carrier_size;
```
[the remark this answers is not in the document]
Comment (koffes) [reply include/audio_defines.h:61]: bps = bits per second.

File: include/audio_defines.h [hunk 7]
```diff
@@ -20,4 +23,91 @@ enum audio_channel {
...
+
+	/* A flag indicating if the samples are interleaved or not.
+	 * False = not interleaved, true = interleaved.
+	 */
```
[the remark this answers is not in the document]
Comment (koffes) [reply include/audio_defines.h:65]: Graham to comment, but I really don't see that we will ever need interleaved. I have removed it.

File: include/audio_defines.h [hunk 8]
```diff
@@ -20,4 +23,91 @@ enum audio_channel {
...
+	uint32_t reference_ts_us;
+
+	/* The timestamp for when the data was received. */
+	uint32_t data_rx_ts_us;
+
+	/* A Boolean flag to indicate this data has errors
+	 * (true = bad, false = good).
+	 * Note: Timestamps may still be valid even though this flag is set.
+	 */
+	bool bad;
```
[the remark this answers is not in the document]
Comment (koffes) [reply include/audio_defines.h:89]: I can rename, but from the host this is called `bad_frame`

[the remark this answers is not in the document]
Comment (koffes) [reply include/audio_defines.h:89]: Renamed to `bad_data`

File: include/audio_defines.h [hunk 9]
```diff
@@ -11,6 +11,9 @@
 #ifndef _AUDIO_DEFINES_H_
 #define _AUDIO_DEFINES_H_
```
[the remark this answers is not in the document]
Comment (asbjornsabo) [reply include/audio_defines.h:12]: As far as I can see, the only real thing that was in this file before was a single enum with the left and right audio channels (lines 17 - 25 after the commit).

I can agree that renaming the file may be outside the scope of this PR, but it should be done.

File: include/audio_defines.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (asbjornsabo) [reply include/audio_defines.h:47]: As long as we are limiting it somewhere else ...
But I guess that typically the sampling frequency values defined in the LE Audio host will be used.

File: include/audio_defines.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (asbjornsabo) [reply include/audio_defines.h:58]: True.

But what about the "carrier_size", which the comment mainly was about?

File: include/audio_defines.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (asbjornsabo) [reply include/audio_defines.h:42]: It is (implicitly) there, in the audio_data.
There is now the possibility that the metadata will say one thing about how much data this is (in seconds), while the data will say something else (in seconds).

File: include/audio_defines.h [hunk 10]
```diff
@@ -20,4 +23,91 @@ enum audio_channel {
...
+ * This unit can be used anywhere, so it may be an audio block, a frame or something else.
+ * It may contain encoded or raw data, as well as a single or multiple channels.
+ */
+struct audio_data {
+	/* A pointer to the raw or coded data (e.g., PCM, LC3, etc.) buffer. */
+	void *data;
+
+	/* The size in bytes of the data buffer.
+	 * To get the size of each channel, this value must be divided by the number of
+	 * used channels. Metadata is not included in this figure.
```
[the remark this answers is not in the document]
Comment (asbjornsabo) [reply include/audio_defines.h:94]: Agreed to remove this added comment again.

File: include/audio_defines.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (asbjornsabo) [reply include/audio_defines.h:42]: Fixed by adjusting comment in audio_data

File: include/audio_defines.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (koffes) [reply include/audio_defines.h]: Changed after our discussion.

**Review by asbjornsabo — APPROVED** [review]
LGTM

**Review by alexsven — APPROVED** [review]

**Pull request merged**